File: csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private readonly int _speed;
    private readonly int _drain;
    private int _distanceDriven;
    private int _battery;
 
    public RemoteControlCar(int speed, int drain)
    {
        _speed = speed;
        _drain = drain;
        _distanceDriven = 0;
        _battery = 100;
    }

    public bool BatteryDrained() => _battery < _drain;

    public int DistanceDriven() => _distanceDriven;

    public void Drive()
    {
        if (_battery < _drain)
        {
            return;
        }
        _distanceDriven += _speed;
        _battery -= _drain;
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);

    public int Reichweite() => _battery / _drain * _speed;
}

class RaceTrack
{
    private readonly int _distance;

    public RaceTrack(int distance)
    {
        _distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car) => _distance <= car.Reichweite();
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
  private int battery = 100;
  private int distance = 0;
  private int speed;
  private int drain;

  NeedForSpeed(int speed, int batteryDrain) {
    this.speed = speed;
    this.drain = batteryDrain;
  }

  public boolean batteryDrained() {
    return battery < drain;
  }

  public int distanceDriven() {
    return distance;
  }

  public void drive() {
    if (batteryDrained()) {
      return;
    }
    battery -= drain;
    distance += speed;
  }

  public int maxDistance() {
    return 100 / drain * speed;
  }

  public static NeedForSpeed nitro() {
    return new NeedForSpeed(50, 4);
  }
}

class RaceTrack {
  private int distance;

  RaceTrack(int distance) {
    this.distance = distance;
  }

  public boolean canFinishRace(NeedForSpeed car) {
    return distance <= car.maxDistance();
  }
}
